include/component.hxx

```diff
@@ -10,43 +10,159 @@
 ///   (std::string name, Options &options, Solver *solver)
-/// 
+///
 struct Component {
+  /// Initialise the `state_variable_acceess` permissions. Note that
+  /// `{all_species}` in any variable names will be replaced with the
+  /// names of all species being simulated (by claling
+  /// `declareAllSpecies()`, which is done after all components are
+  /// created by a ComponentSchedular).
+  Component(Permissions&& access_permissions)
+      : state_variable_access(access_permissions) {}
+
   virtual ~Component() {}
 
-  /// Modify the given simulation state
-  /// All components must implement this function
-  virtual void transform(Options &state) = 0;
-  
+  /// Return a list of names/types of other components needed by this
+  /// component. All configurations for these components will take the
+  /// default value, unless set in the input file.
+  virtual std::vector<ComponentInformation> additionalComponents() { return {}; }
+
+  /// Modify the given simulation state. This method will wrap the
+  /// state in a GuardedOptions object and pass that to the private
+  /// implementation of transform provided by each component.
+  void transform(Options& state);
+
   /// Use the final simulation state to update internal state
   /// (e.g. time derivatives)
-  virtual void finally(const Options &UNUSED(state)) { }
+  virtual void finally(const Options& UNUSED(state)) {}
 
   /// Add extra fields for output, or set attributes e.g docstrings
-  virtual void outputVars(Options &UNUSED(state)) { }
+  virtual void outputVars(Options& UNUSED(state)) {}
 
   /// Add extra fields to restart files
-  virtual void restartVars(Options &UNUSED(state)) { }
+  virtual void restartVars(Options& UNUSED(state)) {}
 
   /// Preconditioning
-  virtual void precon(const Options &UNUSED(state), BoutReal UNUSED(gamma)) { }
-  
+  virtual void precon(const Options& UNUSED(state), BoutReal UNUSED(gamma)) {}
+
   /// Create a Component
   ///
   /// @param type     The name of the component type to create (e.g. "evolve_density")
   /// @param name     The species/name for this instance.
   /// @param options  Component settings: options[name] are specific to this component
   /// @param solver   Time-integration solver
-  static std::unique_ptr<Component> create(const std::string &type, // The type to create
-                                           const std::string &name, // The species/name for this instance
-                                           Options &options,  // Component settings: options[name] are specific to this component
-                                           Solver *solver); // Time integration solver
+  static std::unique_ptr<Component>
+  create(const std::string& type, // The type to create
+         const std::string& name, // The species/name for this instance
+         Options&
+             options, // Component settings: options[name] are specific to this component
+         Solver* solver); // Time integration solver
+
```

Comment:
Formatting's going funky due to the inline comments -- I know you've not touched them, just applied formatting, but may as well remove them given we have the same info immediately above!